storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3109,10 +3110,1754 @@ buf_block_t *fsp_get_latched_xdes_page(
+
+    if (n_used == 1)
+    {
+      flst::remove_complete(m_iblock,
+                            uint16_t(m_ioffset + FSEG_NOT_FULL),
+                            m_old_descr, m_mtr);
+
+      xdes_set_state(*m_old_xdes, m_old_descr, XDES_FREE, m_mtr);
+      flst::append_complete(m_header_block,
+                            FSP_HEADER_OFFSET + FSP_FREE,
+                            m_old_xdes, m_old_xoffset, m_mtr);
+      fil_system.sys_space->free_len++;
+    }
+  }
+
+public:
+  pageOperator(buf_block_t *header_block, buf_block_t *iblock,
+               fseg_inode_t *inode,
+               uint32_t extent_size, uint32_t old_page_no,
+	       mtr_t *mtr) :
+  m_header_block(header_block),
+  m_iblock(iblock), m_inode(inode), m_extent_size(extent_size),
+  m_old_page_no(old_page_no), m_mtr(mtr)
+  {
+    m_free_limit= mach_read_from_4(FSP_HEADER_OFFSET + FSP_FREE_LIMIT
+                                   + m_header_block->page.frame);
+    m_seg_id= mach_read_from_8(m_inode + FSEG_ID);
+  }
+
+  ~pageOperator()
+  {
+    for (uint32_t i= 0; i < 8; i++)
+    {
+      if (m_old_pages[i])
+        buf_block_free(m_old_pages[i]);
+    }
+  }
+
+  /** Get allocated new block */
+  buf_block_t* get_new_block() const noexcept { return m_new_block; }
+
+  /** Prepare the new page allocation from the new given extent
+  @param new_extent starting page of new extent
+  @param segment    segment allocation
+  @return error code */
+  dberr_t prepare_new_page(uint32_t new_extent, bool segment)
+  {
+    dberr_t err= DB_SUCCESS;
+    uint32_t size= mach_read_from_4(FSP_HEADER_OFFSET + FSP_SIZE
+                                    + m_header_block->page.frame);
+    if (new_extent >= size || new_extent >= m_free_limit)
+      return DB_CORRUPTION;
+
+    uint32_t new_descr_page_no= xdes_calc_descriptor_page(0, new_extent);
+    m_new_xdes= fsp_get_latched_page(page_id_t{0, new_descr_page_no},
+                                     m_mtr, &err);
+    if (!m_new_xdes)
+      return err;
+
+    m_ioffset= uint16_t(m_inode - m_iblock->page.frame);
+    m_need_segment= segment;
+    m_xoffset= uint16_t(xdes_calc_descriptor_index(0, new_extent) * XDES_SIZE
+                          + XDES_ARR_OFFSET + XDES_FLST_NODE);
+    m_new_descr= m_new_xdes->page.frame + m_xoffset - XDES_FLST_NODE;
+    m_new_state= uint32_t(xdes_get_state(m_new_descr));
+    uint32_t new_page= 0;
+
+    /* Allocate the new extent and initialize the extent state
+    with XDES_FSEG/XDES_FREE_FRAG */
+    if (m_new_state == XDES_FREE)
+    {
+      if (segment) err= initialize_segment_prepare();
+      else err= initialize_free_frag_prepare();
+
+      if (err) return err;
+new_page:
+      new_page= xdes_find_free(m_new_descr);
+      if (new_page == FIL_NULL)
+        return DB_CORRUPTION;
+
+      new_page += new_extent;
+      m_new_block= fsp_get_latched_page(page_id_t{0, new_page},
+                                        m_mtr, &err);
+      if (err == DB_SUCCESS)
```

Comment:
It would be more efficient to check `if (m_new_block != nullptr)`. It will readily be available in the return value register.
Before  the assignment, or at the start of this function, we should assert `ut_ad(!m_new_block)`.